<a href="https://colab.research.google.com/github/eughappy/HomeWork/blob/RNN/hw14.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
from keras.datasets import imdb
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Embedding, SimpleRNN, LSTM, Dense, Bidirectional

max_features = 10000
maxlen = 500
batch_size = 32

In [3]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)

17464789/17464789 [==============================] - 0s 0us/step


In [6]:
model = Sequential()
model.add(Embedding(max_features, 32, input_length=maxlen))
model.add(SimpleRNN(32))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, validation_data=(x_test,y_test), epochs=5, batch_size=64)
score, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print(f'Test accuracy of RNN model: {acc:.4f}\n')

Epoch 1/5
391/391 [==============================] - 57s 144ms/step - loss: 0.5955 - accuracy: 0.6571 - val_loss: 0.4180 - val_accuracy: 0.8207
Epoch 2/5
391/391 [==============================] - 61s 157ms/step - loss: 0.3753 - accuracy: 0.8417 - val_loss: 0.4298 - val_accuracy: 0.8114
Epoch 3/5
391/391 [==============================] - 62s 158ms/step - loss: 0.2339 - accuracy: 0.9088 - val_loss: 0.4370 - val_accuracy: 0.8380
Epoch 4/5
391/391 [==============================] - 57s 146ms/step - loss: 0.1145 - accuracy: 0.9618 - val_loss: 0.4719 - val_accuracy: 0.8301
Epoch 5/5
782/782 [==============================] - 19s 25ms/step - loss: 0.5806 - accuracy: 0.8130
Test accuracy of RNN model: 0.8130



In [7]:
model = Sequential()
model.add(Embedding(max_features, 32, input_length=maxlen))
model.add(LSTM(32))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, validation_data=(x_test,y_test), epochs=5, batch_size=64)
score, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print(f'Test accuracy of LTSM model: {acc:.4f}\n')

Epoch 1/5
391/391 [==============================] - 111s 279ms/step - loss: 0.4083 - accuracy: 0.8118 - val_loss: 0.3013 - val_accuracy: 0.8745
Epoch 2/5
391/391 [==============================] - 134s 343ms/step - loss: 0.2337 - accuracy: 0.9108 - val_loss: 0.3217 - val_accuracy: 0.8671
Epoch 3/5
391/391 [==============================] - 109s 278ms/step - loss: 0.1783 - accuracy: 0.9333 - val_loss: 0.3581 - val_accuracy: 0.8713
Epoch 4/5
391/391 [==============================] - 131s 334ms/step - loss: 0.1529 - accuracy: 0.9444 - val_loss: 0.3735 - val_accuracy: 0.8567
Epoch 5/5
782/782 [==============================] - 34s 43ms/step - loss: 0.3561 - accuracy: 0.8593
Test accuracy of LTSM model: 0.8593



In [8]:
model = Sequential()
model.add(Embedding(max_features, 32, input_length=maxlen))
model.add(Bidirectional(LSTM(32)))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, validation_data=(x_test,y_test), epochs=5, batch_size=64)
score, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print(f'Test accuracy of biderectional LTSM model: {acc:.4f}\n')

Epoch 1/5
391/391 [==============================] - 202s 505ms/step - loss: 0.4677 - accuracy: 0.7740 - val_loss: 0.3684 - val_accuracy: 0.8540
Epoch 2/5
391/391 [==============================] - 199s 508ms/step - loss: 0.2666 - accuracy: 0.8960 - val_loss: 0.3337 - val_accuracy: 0.8658
Epoch 3/5
391/391 [==============================] - 201s 514ms/step - loss: 0.1896 - accuracy: 0.9310 - val_loss: 0.3114 - val_accuracy: 0.8730
Epoch 4/5
391/391 [==============================] - 200s 512ms/step - loss: 0.1473 - accuracy: 0.9486 - val_loss: 0.3478 - val_accuracy: 0.8634
Epoch 5/5
782/782 [==============================] - 57s 73ms/step - loss: 0.3826 - accuracy: 0.8648
Test accuracy of biderectional LTSM model: 0.8648



In [9]:
model = Sequential()
model.add(Embedding(max_features, 32, input_length=maxlen))
model.add(LSTM(32, return_sequences=True))
model.add(LSTM(32))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, validation_data=(x_test,y_test), epochs=5, batch_size=64)
score, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print(f'Test accuracy of deep LTSM model: {acc:.4f}\n')

Epoch 1/5
391/391 [==============================] - 268s 674ms/step - loss: 0.4088 - accuracy: 0.8074 - val_loss: 0.3262 - val_accuracy: 0.8661
Epoch 2/5
391/391 [==============================] - 225s 576ms/step - loss: 0.2285 - accuracy: 0.9141 - val_loss: 0.3252 - val_accuracy: 0.8615
Epoch 3/5
391/391 [==============================] - 263s 673ms/step - loss: 0.1676 - accuracy: 0.9401 - val_loss: 0.3205 - val_accuracy: 0.8697
Epoch 4/5
391/391 [==============================] - 263s 673ms/step - loss: 0.1283 - accuracy: 0.9549 - val_loss: 0.4596 - val_accuracy: 0.8548
Epoch 5/5
782/782 [==============================] - 71s 90ms/step - loss: 0.4675 - accuracy: 0.8627
Test accuracy of deep LTSM model: 0.8627



Найкращі результати показала BRNN з LTSM шаром. Вона має точність 0.8648 та втрати 0.3826. Вона гарна, отримує доступ до ознак вхідних даних, як минулих, так і майбутніх, а на наших даних це виявилося найбільш ефективним методом.
На другомі місці виявилась глипока рекурентна мережа, з точністю 0.8627 та втратами 0.4675, що свідчить, що модель не дуже точна. На третьому місці за точністю є звичайна GRU модель з шаром LTSM з точністю 0.8593 та втратами 0.3561, а на останньому місці звиччайна RNN ,яка має точність 0.8130 та втратами 0.5806. На цьому наборі даних, як ми бачимо, запам'ятовування минулих ознак показало свою ефективність. Модель з багатьма шарами не виявилася ефективною, проте модель GRU звичайна, якщо її підлаштувати під задані дані, покаже кращий результат. Загалом GRU моделі мають більш кращу точність для роботи з послідовностями.